File: go/src/p030/p030.go
package main

import (
    "fmt"
)

func main() {

    m := make(map[int]int)
    for i := 0; i < 10; i++ {
        m[i] = i * i * i * i * i
    }

    total := 0
    for i := 2; i < 1700000; i++ {
        if i == f(i, m) {
            total += i
        }
    }
    fmt.Println(total)
}

func f(n int, m map[int]int) (int) {
    sum := 0
    for ; n > 0; n /= 10 {
        sum += m[n % 10]
    }
    return sum
}

File: c/p030.c
#include <stdio.h>
#include <stdlib.h>

#include "util.h"

int main(int argc, char** argv) {

    int digits[10];
    for (int i = 0; i < 10; ++i) {
        digits[i] = i * i * i * i * i;
    }

    int sum = 0;
    for (int i = 1000; i < 1000000; ++i) {
        int total = 0;
        for (int n = i; n > 0; n /= 10) {
            total += digits[n % 10];
        }
        if (total == i) {
            sum += total;
        }
    }

    printf("%d\n", sum);
}
